packages/eslint-plugin/src/rules/strict-comparisons.ts

```diff
@@ -0,0 +1,314 @@
+    const typeNames = {
+      [TypeKind.Any]: 'any',
+      [TypeKind.Number]: 'number',
+      [TypeKind.Enum]: 'enum',
+      [TypeKind.String]: 'string',
+      [TypeKind.Boolean]: 'boolean',
+      [TypeKind.Null]: 'null',
+      [TypeKind.Undefined]: 'undefined',
+      [TypeKind.Object]: 'object',
+    };
+
+    /**
+     * Get TypeKinds of a typescript type
+     * @param type
+     */
+    function getKinds(type: ts.Type): TypeKind[] {
+      return isUnionType(type)
+        ? Array.from(new Set(type.types.map(getKind)))
+        : [getKind(type)];
+    }
+
+    /**
+     * Get TypeKind of a typescript type
+     * @param type
+     */
+
+    function getKind(type: ts.Type): TypeKind {
+      if (is(ts.TypeFlags.String | ts.TypeFlags.StringLiteral)) {
+        return TypeKind.String;
+      } else if (is(ts.TypeFlags.Number | ts.TypeFlags.NumberLiteral)) {
+        return TypeKind.Number;
+      } else if (is(ts.TypeFlags.BooleanLike)) {
+        return TypeKind.Boolean;
+      } else if (is(ts.TypeFlags.Null)) {
+        return TypeKind.Null;
+      } else if (is(ts.TypeFlags.Undefined)) {
+        return TypeKind.Undefined;
+      } else if (is(ts.TypeFlags.Any)) {
+        return TypeKind.Any;
+      } else {
+        return TypeKind.Object;
+      }
+
+      function is(flags: ts.TypeFlags) {
+        return isTypeFlagSet(type, flags);
+      }
+    }
+
+    /**
+     * Check if a specific TypeKind is present in an array
+     * @param typesLeft array of TypeKinds
+     * @param typesRight TypeKind to check
+     */
+    function getStrictestComparableType(
+      typesLeft: TypeKind[],
+      typesRight: TypeKind[],
+    ): TypeKind | undefined {
+      const overlappingTypes = typesLeft.filter(
+        type => typesRight.indexOf(type) >= 0,
+      );
+
+      if (overlappingTypes.length > 0) {
+        return getStrictestKind(overlappingTypes);
+      } else {
+        // In case one of the types is "any", get the strictest type of the other array
+        if (arrayContainsKind(typesLeft, [TypeKind.Any])) {
+          return getStrictestKind(typesRight);
+        }
+        if (arrayContainsKind(typesRight, [TypeKind.Any])) {
+          return getStrictestKind(typesLeft);
+        }
+
+        // In case one array contains Null or Undefined and the other an Object, return Object
+        if (
+          (arrayContainsKind(typesLeft, [TypeKind.Null, TypeKind.Undefined]) &&
+            arrayContainsKind(typesRight, [TypeKind.Object])) ||
+          (arrayContainsKind(typesRight, [TypeKind.Null, TypeKind.Undefined]) &&
+            arrayContainsKind(typesLeft, [TypeKind.Object]))
+        ) {
+          return TypeKind.Object;
+        }
+        return undefined;
+      }
+    }
+
+    /**
+     * Check if a specific TypeKind is present in an array
+     * @param types array of TypeKinds
+     * @param typeToCheck TypeKind to check
+     */
+    function arrayContainsKind(
+      types: TypeKind[],
+      typeToCheck: TypeKind[],
```

Comment:
bit of a nit, but could be a more ergonomic api if you use the rest operator - would save you manually constructing the arrays for each of your calls.
```suggestion
      ...typeToCheck: TypeKind[],
```